### PR TITLE
Add support for evaluator functions to parser

### PR DESCRIPTION
**Pull request by tohrnii**
Partly addressing #124

### CODE REVIEW

File: parser/src/ast/evaluator_function.rs [hunk 1]
```diff
@@ -0,0 +1,73 @@
+use super::{Identifier, IntegrityStmt, TraceCols};
+
+/// Evaluator functions take description of the main and auxiliary traces as input, and returns
+/// integrity constraints that are enforced on those trace columns.
```
Comment (grjte) [inline parser/src/ast/evaluator_function.rs:4]: As noted in #122 and #213, we may need to allow passing periodic columns and random values.

Also, I would rephrase the end of this from "and returns..." to "and enforces integrity constraints on those trace columns."

File: parser/src/ast/evaluator_function.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+    /// Returns the auxiliary trace of the evaluator function.
+    pub fn aux_trace(&self) -> &Vec<TraceCols> {
+        &self.aux_trace
+    }
+
+    /// Returns the integrity statements of the evaluator function.
+    pub fn integrity_stmts(&self) -> &[IntegrityStmt] {
+        &self.integrity_stmts
+    }
```
Comment (grjte) [inline parser/src/ast/evaluator_function.rs:47]: I think we'll want to add a new `into_parts` method after this so that we can consume each `EvaluatorFunction` and add it to the symbol table when consuming the `AST` during the IR step. However, if you prefer to add this during the PR for the IR changes that's ok too

File: parser/src/ast/evaluator_function.rs [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+/// Evaluator function call is used to invoke an evaluator function. It takes a list of trace
+/// columns.
+#[derive(Debug, Eq, PartialEq)]
+pub struct EvaluatorFunctionCall {
+    name: Identifier,
+    args: Vec<Vec<TraceCols>>,
+}
```
Comment (grjte) [inline parser/src/ast/evaluator_function.rs:56]: similar comment here - I think we may need to revisit this, if we're allowing args to be identifiers as well (e.g. for periodic columns or random values)

File: parser/src/ast/evaluator_function.rs [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+    /// Returns the name of the evaluator function.
+    pub fn name(&self) -> &str {
+        self.name.name()
+    }
+
+    /// Returns the arguments of the evaluator function.
+    pub fn args(&self) -> &Vec<Vec<TraceCols>> {
+        &self.args
+    }
```
Comment (grjte) [inline parser/src/ast/evaluator_function.rs:72]: similar comment here about adding `into_parts`

File: parser/src/ast/mod.rs [hunk 5]
```diff
@@ -36,6 +39,8 @@ pub struct Source(pub Vec<SourceSection>);
 ///   elements specified. The array length is expected to be a power of 2, but this is not checked
 ///   during parsing.
 /// - RandomValues: Random Values represent the randomness sent by the Verifier.
+/// - EvaluatorFunction: Evaluator Functions take description of the main and auxiliary traces as
+///   input, and returns integrity constraints that are enforced on those trace columns.
```
Comment (grjte) [inline parser/src/ast/mod.rs:43]: 1. nit: there's a typos "take description" -> "take descriptions"
2. I would adjust the end of it to "and enforce integrity constraints on those trace columns."
3. This may need to change if we allow periodic columns or random values as inputs

File: parser/src/parser/tests/integrity_constraints.rs [hunk 6]
```diff
@@ -232,7 +232,8 @@ fn err_missing_integrity_constraint() {
         let b = [a, 2 * a]
         let c = [[a - 1, a^2], [b[0], b[1]]]";
     let error = Error::ParseError(ParseError::MissingIntegrityConstraint(
-        "Declaration of at least one integrity constraint is required".to_string(),
+        "Declaration of at least one integrity constraint or evaluator function call is required"
+            .to_string(),
```
Comment (grjte) [inline parser/src/parser/tests/integrity_constraints.rs:236]: I probably wouldn't change this. In my mind, an evaluator function call is just a special type of constraint. It's still prefixed by the `enf` keyword that we use for constraints, so I don't think the user should need to think about it as a separate thing.

File: parser/src/parser/grammar.lalrpop [hunk 7]
```diff
@@ -155,6 +155,67 @@ RandElem: RandBinding = {
     <name: Identifier> <size: Size> => RandBinding::new(name, size)
 }
 
+// EVALUATOR FUNCTIONS
+// ================================================================================================
+
+EvaluatorFunction: EvaluatorFunction = {
+    "ev" <evaluator_fn_name: Identifier>
+        "(" <main_cols: MainCols?> <cols_separator: (",")?> <aux_cols: AuxCols?> ")" ":"
```
Comment (grjte) [inline parser/src/parser/grammar.lalrpop:163]: Could we simplify this to the following? Then we could remove the error checking for the `cols_separator` to simplify the error handling below as well.

```
"(" <main_cols: MainCols?> "," <aux_cols: AuxCols?> ")" ":"
```

File: parser/src/parser/grammar.lalrpop [hunk 8]
```diff
@@ -240,14 +301,14 @@ IntegrityConstraints: Vec<IntegrityStmt> = {
...
-            IntegrityStmt::Constraint(_) => true,
+            IntegrityStmt::Constraint(_) | IntegrityStmt::EvaluatorFunctionCall(_) => true,
             _ => false,
         });
         if !integrity_constraints_exist {
             return Err(ParseError::User {
                 error: Error::ParseError(
                     MissingIntegrityConstraint(
-                        "Declaration of at least one integrity constraint is required".to_string()
+                        "Declaration of at least one integrity constraint or evaluator function call is required".to_string()
```
Comment (grjte) [inline parser/src/parser/grammar.lalrpop:311]: This looks like the same error message as above. It would be good to extract this so it's not duplicated. This could be improved as part of #188 though

**Review by grjte — CHANGES_REQUESTED** [review]
@tohrnii I think this looks great. I left a few minor comments inline and also flagged a few places that might be affected by the proposed modularity changes mentioned in #122 and #213.

However, I think those changes (if we decide to make them) could be addressed in a subsequent PR, so we could still merge this one and implement the IR changes for this initial evaluator function support first (once the other minor comments are addressed)

File: parser/src/parser/grammar.lalrpop (hunk 7 above)
Comment (tohrnii) [reply parser/src/parser/grammar.lalrpop:163]: Hanging comma seems a little weird to me personally. (e.g. `(main: [a, b, c], )` or `(, aux: [a, b, c])`. But if you think it's worth it to keep the grammar cleaner, I can change it.

File: parser/src/parser/tests/integrity_constraints.rs (hunk 6 above)
Comment (tohrnii) [reply parser/src/parser/tests/integrity_constraints.rs]: Makes sense. Thanks.

File: parser/src/ast/evaluator_function.rs [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    /// Returns the integrity statements of the evaluator function.
+    pub fn integrity_stmts(&self) -> &[IntegrityStmt] {
+        &self.integrity_stmts
+    }
+
+    /// Returns the name, main trace columns, auxiliary trace columns, and integrity statements
+    /// of the evaluator function.
+    pub fn into_parts(self) -> (String, Vec<TraceCols>, Vec<TraceCols>, Vec<IntegrityStmt>) {
+        (
+            self.name.into_name(),
```
Comment (grjte) [inline parser/src/ast/evaluator_function.rs:53]: I would return the Identifier here for consistency, since that's what's been done elsewhere

File: parser/src/ast/evaluator_function.rs [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+
+    /// Returns the arguments of the evaluator function.
+    pub fn args(&self) -> &Vec<Vec<TraceCols>> {
+        &self.args
+    }
+
+    /// Returns the name and arguments of the evaluator function.
+    pub fn into_parts(self) -> (String, Vec<Vec<TraceCols>>) {
+        (self.name.into_name(), self.args)
+    }
```
Comment (grjte) [inline parser/src/ast/evaluator_function.rs:88]: I would return the `Identifier` here instead of a `String` for consistency with other places

**Review by grjte — CHANGES_REQUESTED** [review]
@tohrnii thanks! I left 2 minor comments and a suggestion for simplifying the grammar

File: parser/src/parser/grammar.lalrpop (hunk 7 above)
Comment (grjte) [reply parser/src/parser/grammar.lalrpop:161]: Another benefit of adjusting in this way is that you can use `IntegrityStmts` for the `integrity_constraints` parsing so that the error handling isn't duplicated

File: parser/src/ast/evaluator_function.rs (hunk 9 above)
Comment (tohrnii) [reply parser/src/ast/evaluator_function.rs:53]: I think `String` is returned elsewhere as well?

File: parser/src/parser/grammar.lalrpop (hunk 7 above)
Comment (tohrnii) [reply parser/src/parser/grammar.lalrpop:161]: Thank you! This is much cleaner. Fixed.

File: parser/src/ast/evaluator_function.rs (hunk 9 above)
Comment (grjte) [reply parser/src/ast/evaluator_function.rs:53]: Ah yes, it looks like I misremembered this and you're completely right, sorry. I went through so many iterations of things over the course of the refactor that it seems to just be a weird jumble in my head at this point

File: parser/src/parser/tests/evaluator_functions.rs [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+            Identifier("advance_clock".to_string()),
+            vec![vec![TraceCols::new(Identifier("clk".to_string()), 1)]],
+        )),
+    ])]);
+
+    build_parse_test!(source).expect_ast(expected);
+}
+
+#[test]
+fn ev_fun_call_inside_ev_fn() {
```
Comment (grjte) [inline parser/src/parser/tests/evaluator_functions.rs:96]: nit: I would probably do `ev_fn_call_inside_ev_fn` for consistency with our other names

**Review by grjte — APPROVED** [review]
Looks good, thanks @tohrnii! I left one optional nit inline
